src/console/man/om_create.1
." Text automatically generated by txt2man
.TH om_create  "March 25, 2008" "" ""
.SH NAME
\fBom_create \fP- create a distribution model using the openModeller framework
\fB
.SH SYNOPSIS
.nf
.fam C
\fBom_create\fP [-] \fIv\fP \fB--version\fP | \fIr\fP \fB--xml-req\fP=XML_REQUEST_FILE \fIm\fP \fB--model-file\fP=FILE [ \fB--log-level\fP=LEVEL ] [ \fB--log-file\fP=FILE ] [ \fB--prog-file\fP=FILE ]
.fam T
.fi
.SH DESCRIPTION
\fBom_create\fP is a command line tool to generate distribution models. The main input is an XML file containing a model request according to the ModelParameters element definition in http://openmodeller.cria.org.br/xml/1.0/openModeller.xsd (see also model_request.xml in the openModeller examples directory). The distribution model generated by openModeller will be stored in another file specified as a parameter. This file is also an XML file, but now following the SerializedModel element definition in http://openmodeller.cria.org.br/xml/1.0/openModeller.xsd. Please note that each algorithm has its own way to represent models. To generate distribution maps, use om_project (you will need a serialized model as a parameter).
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fIr\fP, \fB--xml-req\fP
File containing a model request in XML.
.TP
.B
-\fIm\fP, \fB--model-file\fP
File where the generated model will be stored.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.TP
.B
\fB--log-file\fP
File to store log information.
.TP
.B
\fB--prog-file\fP
File to store progress.
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 0.5.3 (2008-03-26)
--------------------------
SVN revision: 4209

	* Fixed bug when GDAL failed to read a raster row (in which case the row used to have zero values). Now the row is filled with nodata values.
	* XML request for model creation now supports additional options to filter occurrences (using spatiallyUnique or environmentallyUnique sampler functions).
	* Changed nodata value of the default raster type (ByteHFA) to 101.
	* Masks must now use nodata to indicate masked areas (zeros will not work anymore).
	* om_testmodel now generates pseudo-absence points when there are no absences to be tested.
	* Display confusion matrix cell values in om_console and om_testmodel.
	* Renamed getLayerFilename to getLayerPath and getMaskFilename to getMaskPath in EnvironmentPtr class.
	* Refactored om_pseudo, om_create and om_project to use the getopts command line library.
	* Created man pages for om_pseudo, om_create and om_project.
	* New parameter in om_pseudo to speficy the proportion of absence points to be generated.
	* New algorithm AquaMaps (for marine organisms).
	* Removed algorithms minimum distance and distance to average since EnvironmentalDistance provides the same functionality.
	* "type" property of AlgParamMetadata changed from char * to a new enumeration called AlgParamDatatype. Values can be Integer, Real and String.
	* Fixed bug in the environmental distance algorithm: probabilities were negative for points whose distance to the nearest point was beyond the maximum allowed distance. 
	* New classes to perform jackknife and chi-square in the environmental layers.
	* Updated TerraLib drivers for compatibility with TerraLib version 3.2.0.

Release 0.5.2 (2007-10-24)
--------------------------
SVN revision: 3806

	* Fixed compilation issues under Windows.
	* Included new command line program to generate pseudo occurrences.
	* Minor improvements in console tools (absences are now displayed in om_viewer and om_niche).
	* Code clean up.

Release 0.5.1 (2007-08-30)
--------------------------
SVN revision: 3661

	* Fixed MSVC compilation problems.
	* Fixed bug in deserialization of the new GARP algorithm (OM GARP).
	* Fixed crash in one-class SVM when input points contained absences.
	* Fixed bug in deserialization of environmental distance algorithm with Mahalanobis distance.
	* Implemented serialization/deserialization for the new GARP Best Subsets (OM GARP).
	* New algorithm "Envelope Score".
	* Fixed bug in the pseudo-absence generation of the SVM algorithm when no absences were passed as a parameter.

Release 0.5 (2007-08-15)
------------------------
SVN revision: 3527

	* New algorithm "Support Vector Machines" (C-SVC, nu-SVC and one-class SVM).
	* Added support for multiple normalization techniques (two implementations are available: ScaleNormalizer and MeanVarianceNormalizer).
	* New method to cancel jobs (model creation or model projection).
	* Sample serialization is now based on the original (unnormalized) environment values.
	* New infrastructure for unit tests using cxxtest.

Release 0.4.2 (2007-05-08)
--------------------------

	* Included ROC curve as part of model statistics.
	* Added metric Chebyshev in the Environmental Distance algorithm.
	* Log object is now a singleton.
	* Minor bugfixes.

Release 0.4.1 (2007-03-12)
--------------------------

	* New algorithm "Environmental Distance".
	* Changed OccurrencesReader to handle presence and absence points separately.
	* Changed occurrences file format (now TAB-delimited with each point having a unique identifier).
	* Initial support to cmake (should work on GNU/Linux and Mac OSX).

Release 0.4 (2006-12-04)
------------------------

	* Integration with TerraLib.
	* Removed the algorithm Bioclim distance.
	* Changed the default output map file type to ERDAS Imagine ".img".
	* Refactored SOAP interface which now covers all modelling functionality.
	* Re-enabled the new version of GARP (v3).
	* Several fixes in the SWIG-Python binding.
	* Fixed installation issue with the RPM packages.

Release 0.3.4 (2005-12-14)
--------------------------

	* Fixed bug in model projector that could cause localized distortions in the distribution map (near the grid cell scale).
	* New version of the CSM algorithm with several adjustments in code (Broken Stick cutoff method).
	* New command line tools om_sampledump, om_create, om_project.

Release 0.3.3 (2005-09-01)
--------------------------

	* Implemented serialization mechanism for the Best Subsets Procedure.
	* Moved the createModel logic and ModelCommand object into Algorithm.
	* Fixed issues related to the Mac OS X build (for both 10.3 and 10.4).

Release 0.3.2 (2005-08-11)
--------------------------

	* Fixed bug that was making GARP produce null models (100% omission) most part of the time.

Release 0.3.1 (2005-07-18)
--------------------------

	* Major restructuring of directories, file locations and file names.

Release 0.3 (planned to 2005-05-25, but never happened)
-------------------------------------------------------

	* Reimplemented serialization/deserialization mechanism using generic configuration objects.
	* All available algorithms are serializable.
	* om_console accepts new keywords "Output model" (file name to store the serialized model) and "Input model" to load a serialized model (instead of using "WKT format", "Species file", "Species" and "Map" to generate a model).
	* New framework for test cases is available (depends on the SWIG/Python interface).
	* Moved normalization from the individual raster files to the "environment" object.
	* Implemented Model and Algorithm separation (new interface called Model abstracts the portion of Algorithm used for evaluations).
	* Implemented reference-counting smart pointers for all major objects.
	* Fixed various problems with memory leaks and uninitialized values.
	* Removed CSM Kaiser-Gutman from build.
	* Removed SWIG/Java binding from build.

src/console/om_create.cpp
/*************************************************************************************
 *  A simple command line app to create an openModeller model
 *    -------------------
 *       begin                : November 2005
 *       copyright            : (C) 2005 by T.Sutton, Kevin Ruland, Renato De Giovanni
 *       email                : tim@linfiniti.com
 *************************************************************************************/

/***************************************************************************
 *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
#include <openmodeller/om.hh>
#include <openmodeller/Log.hh>

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"

#include <fstream>   // file I/O for XML
#include <sstream>   // ostringstream datatype
#include <stdio.h>   // file I/O for log
#include <time.h>    // used to limit the number of times that the progress is written to a file
#include <string>    // string library
#include <stdexcept> // try/catch

using namespace std;

/// Main code
int main( int argc, char **argv ) {

  char *args;
  int option;

  struct options opts[] = 
  {
    { 1, "version",    "Display version info",                          "v", 0 },
    { 2, "xml-req",    "Model creation request file in XML",            "r", 1 },
    { 3, "model-file", "File to store the generated model",             "m", 1 },
    { 4, "log-level",  "Set the log level (debug, warn, info, error)", NULL, 1 },
    { 5, "log-file",   "Log file",                                     NULL, 1 },
    { 6, "prog-file",  "File to store model creation progress",        NULL, 1 },
    { 0, NULL,         NULL,                                           NULL, 0 }
  };

  bool passed_params = false;

  std::string log_level("info");
  std::string request_file;
  std::string model_file;
  std::string log_file;
  std::string progress_file;

  while ( ( option = getopts( argc, argv, opts, &args ) ) != 0 ) {

    passed_params = true;

    switch ( option ) {

      // Special Case: Recognize options that we didn't set above.
      case -2: 
        printf( "Unknown option: %s\n", args );
        break;
      // Special Case: getopts() can't allocate memory.
      case -1:
        printf( "Unable to allocate memory from getopts().\n" );
        exit(-1);
        break;
      case 1:
        printf("om_create 0.2.1\n");
        exit(0);
        break;
      case 2:
        request_file = args;
        break;
      case 3:
        model_file = args;
        break;
      case 4:
        log_level = args;
        break;
      case 5:
        log_file = args;
        break;
      case 6:
        progress_file = args;
        break;
      default:
        break;
    }

    // This free() is required since getopts() automagically allocates space 
    // for "args" everytime it's called. */
    free( args );
  }

  if ( ! passed_params ) {

    // Display usage
    getopts_usage( "om_create", opts );
    exit(0);
  }

  // Check parameters

  if ( request_file.empty() ) {

    printf( "Please specify a model creation request file in XML\n");
    exit(-1);
  }

  // Initialize progress data if user wants to track progress
  progress_data prog_data;

  if ( ! progress_file.empty() ) { 

    prog_data.file_name = progress_file;

    time( &prog_data.timestamp );

    prog_data.progress = -3.0;

    // Always create initial file with status "queued" (-1)
    progressFileCallback( -1.0, &prog_data );
  }

  // Log stuff

  Log::Level level_code = getLogLevel( log_level );

  if ( ! log_file.empty() ) {

    Log::instance()->set( level_code, log_file, "" );
  }
  else {
 
    // Just set the level - things will go to stderr
    Log::instance()->setLevel( level_code );
  }

  // Real work

  try {

    // Load algorithms and instantiate controller class
    AlgorithmFactory::searchDefaultDirs();

    OpenModeller om;

    // If user wants to track progress
    if ( ! progress_file.empty() ) { 

      // Set callback to write to a file
      om.setModelCallback( progressFileCallback, &prog_data );
    }
    else if ( ! model_file.empty() ) {

      // Default callback will display progress on screen when a model file was specified
      // (which means the model won't be sent to stdout)
      om.setModelCallback( progressDisplayCallback );
    }

    ConfigurationPtr input = Configuration::readXml( request_file.c_str() );
    om.setModelConfiguration( input );

    om.createModel();

    ConfigurationPtr output = om.getModelConfiguration();

    std::ostringstream model_output;

    Configuration::writeXml( output, model_output );

    std::cerr << flush;

    // Write model output to file, if requested
    if ( ! model_file.empty() ) {

      ofstream file( model_file.c_str() );
      file << model_output.str();
      file.close();
    }
    else {

      // Otherwise send it to stdout
      fprintf( stdout, model_output.str().c_str() );

      std::cout << endl << flush;
    }

    // If user wants to track progress
    if ( ! progress_file.empty() ) { 

      // Check if job was completed
      if ( prog_data.progress != 1 ) {

        // -2 means aborted
        progressFileCallback( -2.0, &prog_data );
      }
    }
  }
  catch ( runtime_error e ) {

    // If user is tracking progress
    if ( ! progress_file.empty() ) { 

      // -2 means aborted
      progressFileCallback( -2.0, &prog_data );
    }

    printf( "om_create aborted: %s\n", e.what() );
  }
}
